capsules/src/net/tlv.rs

```diff
@@ -0,0 +1,1324 @@
+                let context_id_compress = (first_byte & 0b10000) > 0;
+                let context_id = first_byte & 0b1111;
+                let (offset, context_length) = dec_try!(buf, offset; decode_u8);
+                stream_done!(offset,
+                             Some((PrefixSubTlv::SixLoWpanId {
+                                       context_id_compress: context_id_compress,
+                                       context_id: context_id,
+                                       context_length: context_length,
+                                   },
+                                   stable)))
+            }
+            PrefixSubTlvType::NotPresent => stream_done!(offset, None),
+        }
+    }
+}
+
+#[repr(u8)]
+pub enum PrefixSubTlvType {
+    HasRoute = 0,
+    BorderRouter = 2,
+    SixLoWpanId = 3,
+    NotPresent,
+}
+
+impl From<u8> for PrefixSubTlvType {
+    fn from(tlv_type: u8) -> Self {
+        match tlv_type {
+            0 => PrefixSubTlvType::HasRoute,
+            2 => PrefixSubTlvType::BorderRouter,
+            3 => PrefixSubTlvType::SixLoWpanId,
+            _ => PrefixSubTlvType::NotPresent,
+        }
+    }
+}
+
+impl<'a, 'b> From<&'a PrefixSubTlv<'b>> for PrefixSubTlvType {
+    fn from(prefix_sub_tlv: &'a PrefixSubTlv<'b>) -> Self {
+        match *prefix_sub_tlv {
+            PrefixSubTlv::HasRoute(_) => PrefixSubTlvType::HasRoute,
+            PrefixSubTlv::BorderRouter(_) => PrefixSubTlvType::BorderRouter,
+            PrefixSubTlv::SixLoWpanId { .. } => PrefixSubTlvType::SixLoWpanId,
+        }
+    }
+}
+
+pub struct HasRouteTlvValue {
+    r_border_router_16: u16,
+    r_preference: u8,
+}
+
+impl HasRouteTlvValue {
+    pub fn encode(&self, buf: &mut [u8]) -> SResult {
+        stream_len_cond!(buf, 3);
+        let mut offset = enc_consume!(buf, 0; encode_u16, self.r_border_router_16.to_be());
+        let last_byte = ((self.r_preference & 0b11) as u8) << 6;
+        offset = enc_consume!(buf, offset; encode_u8, last_byte);
+        stream_done!(offset)
+    }
+
+    pub fn decode(buf: &[u8]) -> SResult<HasRouteTlvValue> {
+        stream_len_cond!(buf, 3);
+        let (offset, r_border_router_16) = dec_try!(buf; decode_u16);
+        let (offset, last_byte) = dec_try!(buf, offset; decode_u8);
+        let r_preference = last_byte >> 6;
+        stream_done!(offset,
+                     HasRouteTlvValue {
+                         r_border_router_16: r_border_router_16,
+                         r_preference: r_preference,
+                     })
+    }
+}
+
+pub struct BorderRouterTlvValue {
+    // See 15.8.3
+    p_border_router_16: u16,
+    p_bits: u16,
+}
+
+// See 5.18.3
+#[repr(u16)]
+pub enum BorderRouterTlvValueBit {
+    Prf = 0b1100000000000000,
+    P = 0b0010000000000000,
+    S = 0b0001000000000000,
+    D = 0b0000100000000000,
+    C = 0b0000010000000000,
+    R = 0b0000001000000000,
+    O = 0b0000000100000000,
+    N = 0b0000000010000000,
```

Comment:
I figured anyone using this code to manage TLVs will be reading the spec as they do it, and these letters are defined there.

Added comments denoting what each stands for.